compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -140,10 +141,34 @@ impl<'tcx> crate::MirPass<'tcx> for GVN {
             state.visit_basic_block_data(bb, data);
         }
 
-        // For each local that is reused (`y` above), we remove its storage statements do avoid any
-        // difficulty. Those locals are SSA, so should be easy to optimize by LLVM without storage
-        // statements.
-        StorageRemover { tcx, reused_locals: state.reused_locals }.visit_body_preserves_cfg(body);
+        // When emitting storage statements, we want to retain the reused locals' storage statements,
+        // as this enables better optimizations. For each local use location, we mark it for storage removal
+        // only if it might be uninitialized at that point.
+        let storage_to_remove = if tcx.sess.emit_lifetime_markers() {
+            let maybe_uninit = MaybeUninitializedLocals::new()
+                .iterate_to_fixpoint(tcx, body, Some("mir_opt::gvn"))
+                .into_results_cursor(body);
+
+            let mut storage_checker = StorageChecker {
+                reused_locals: &state.reused_locals,
+                storage_to_remove: DenseBitSet::new_empty(body.local_decls.len()),
+                maybe_uninit,
+            };
+
+            for (bb, data) in traversal::reachable(body) {
+                storage_checker.visit_basic_block_data(bb, data);
+            }
+
+            storage_checker.storage_to_remove
+        } else {
+            // Remove the storage statements of all the reused locals.
+            state.reused_locals.clone()
+        };
```

Comment:
This logic is very close to the one in `copy_prop`. Could you merge them into a single `compute_storage_to_remove` function in `rustc_mir_transform::ssa`? It would take a set of relevant locals (copy-prop : `copy_classes[local] != local`, GVN : `reused_locals`) and compute `storage_to_remove`.

Reply:
I might be missing something, but I think that because we run the storage check at different stages (in `copy_prop`, we do this before the replacement) they have to be different: removing the `PlaceContext::MutatingUse` part in GVN results in removing many more storage statements because we end up checking assignment as well.